### PR TITLE
Add `PoolStateMachine.ConnectionGroup` (#425)

### PR DESCRIPTION
**Pull request by fabianfett**

### CODE REVIEW

File: Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift [hunk 1]
```diff
@@ -0,0 +1,634 @@
...
+            self.connections.isEmpty
+        }
+
+        @usableFromInline
+        var canGrow: Bool {
+            self.stats.active < self.maximumConcurrentConnectionHardLimit
+        }
+
+        @usableFromInline
+        var soonAvailable: UInt16 {
```
Comment (gwynne) [inline Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift:111]: This name sounds more like something boolean than a counter. Is there something more specific to call this?

File: Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+            self.stats.connecting -= 1
+            self.stats.backingOff += 1
+
+            return self.connections[index].failedToConnect()
+        }
+
+        @usableFromInline
+        enum BackoffDoneAction {
+            case createConnection(ConnectionRequest, TimerCancellationToken?)
+            case cancelTimers(Max2Sequence<TimerCancellationToken>)
```
Comment (gwynne) [inline Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift:227]: I keep meaning to ask - is `Array` really so non-performant for "collections of 0, 1, or 2 elements" as to make `Max2Sequence` that much worth having?

File: Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift [hunk 3]
```diff
@@ -0,0 +1,634 @@
...
+            }
+        }
+
+        /// Closes the connection at the given index.
+        @inlinable
+        mutating func closeConnection(at index: Int) -> Connection {
+            self.stats.idle -= 1
+            self.stats.closing += 1
+            fatalError()
+//            return self.connections[index].close()
```
Comment (gwynne) [inline Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift:436]: 🤨

File: Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift [hunk 4]
```diff
@@ -0,0 +1,634 @@
...
+
+            self.stats.availableStreams -= closeAction.maxStreams
+
+            return CloseAction(
+                connection: closeAction.connection,
+                timersToCancel: closeAction.cancelTimers
+            )
+        }
+
+        // MARK: Connection failure
```
Comment (gwynne) [inline Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift:476]: Is `connectionClosed()` always a failure state?

File: Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift [hunk 5]
```diff
@@ -0,0 +1,634 @@
...
+                self.stats.idle -= 1
+
+            case .leased:
+                self.stats.leased -= 1
+
+            case .closing:
+                self.stats.closing -= 1
+            }
+
+            let lastIndex = self.connections.endIndex - 1
```
Comment (gwynne) [inline Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift:511]:
```suggestion
            let lastIndex = self.connections.index(before: self.connections.endIndex)
```

File: Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift [hunk 6]
```diff
@@ -0,0 +1,634 @@
...
+        /*private*/ func getConnectionUse(index: Int) -> ConnectionUse {
+            switch index {
+            case 0..<self.minimumConcurrentConnections:
+                return .persisted
+            case self.minimumConcurrentConnections..<self.maximumConcurrentConnectionSoftLimit:
+                return .demand
+            case self.maximumConcurrentConnectionSoftLimit...:
+                return .overflow
+            default:
+                preconditionFailure()
```
Comment (gwynne) [inline Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift:550]: Needs message.

File: Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift [hunk 7]
```diff
@@ -0,0 +1,634 @@
...
+        @inlinable
+        /*private*/ mutating func swapForDeletion(index indexToDelete: Int) -> TimerCancellationToken? {
+            let lastConnectedIndex = self.connections.lastIndex(where: { $0.isConnected })
+
+            if lastConnectedIndex == nil || lastConnectedIndex! < indexToDelete {
+                self.removeO1(indexToDelete)
+                return nil
+            }
+
+            guard let lastConnectedIndex = lastConnectedIndex else { preconditionFailure() }
```
Comment (gwynne) [inline Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift:575]:
```suggestion
            guard let lastConnectedIndex else { preconditionFailure() }
```
Also, with no message you might as well just use `!` for all the difference it makes.

File: Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift [hunk 8]
```diff
@@ -0,0 +1,634 @@
...
+                }
+
+            default:
+                return nil
+            }
+        }
+
+        @inlinable
+        /*private*/ mutating func removeO1(_ indexToDelete: Int) {
+            let lastIndex = self.connections.endIndex - 1
```
Comment (gwynne) [inline Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift:621]:
```suggestion
            let lastIndex = self.connections.index(before: self.connections.endIndex)
```

File: Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift (hunk 2 above)
Comment (fabianfett) [reply Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift:227]: An array of 0 is for free. An array of 2 elements is an allocation. Given how often we run into those situations, I'm happy to save those.

**Review by gwynne — APPROVED** [review]

**Pull request merged**